doc/developer/design/20250226_postgres_style_explain.md

```diff
@@ -0,0 +1,338 @@
+ - We must update `mz_lir_mapping` to use the new vocabulary.
+ - We must update the documentation to explain the new output, ideally
+   using this output everywhere `EXPLAIN` is used.
+
+### Concrete Mapping
+
+| LIR node    | `mz_lir_mapping` node                    | New, Postgres-style syntax              |
+| :---------- | :--------------------------------------- | :-------------------------------------- |
+| `Constant`  | `Constant`                               | `Constant`                              |
+| `Get`       | `Get::PassArrangements l0`               | `Index Scan on l0 using ...`            |
+| `Get`       | `Get::Arrangement l0 (val=...)`          | `Index Lookup on l0 using ...`          |
+| `Get`       | `Get::Collection l0`                     | `Read l0`                               |
+| `Mfp`       | `MapFilterProject`                       | `Map/Filter/Project`                    |
+| `FlatMap`   | `FlatMap`                                | `Flat Map`                              |
+| `Join`      | `Join::Differential`                     | `Differential Join`                     |
+| `Join`      | `Join::Delta`                            | `Delta Join`                            |
+| `Reduce`    | `Reduce::Distinct`                       | `Distinct GroupAggregate`               |
+| `Reduce`    | `Reduce::Accumulable`                    | `Accumulable GroupAggregate`            |
+| `Reduce`    | `Reduce::Hierarchical (monotonic)`       | `Monotonic Hierarchical GroupAggregate` |
+| `Reduce`    | `Reduce::Hierarchical (buckets: ...)`    | `Bucketed Hierarchical GroupAggregate`  |
+| `Reduce`    | `Reduce::Basic`                          | `Non-incremental GroupAggregate`        |
+| `Reduce`    | `Reduce::Collation`                      | `Collated GroupAggregate` (details?)    |
+| `TopK`      | `TopK::MonotonicTop1`                    | `Monotonic Top1`                        |
+| `TopK`      | `TopK::MonotonicTopK`                    | `Monotonic TopK`                        |
+| `TopK`      | `TopK::Basic`                            | `Non-monotonic TopK`                    |
+| `Negate`    | `Negate`                                 | `Negate Diffs`                          |
+| `Threshold` | `Threshold`                              | `Threshold Diffs`                       |
+| `Union`     | `Union`                                  | `Union`                                 |
+| `Union`     | `Union (consolidates output)`            | `Consolidating Union`                   |
+| `ArrangeBy` | `Arrange`                                | `Arrange`                               |
+| `Let`       | `e0 With l1 = e1 ...`                    | `e1 With l1 = e1 ...`                   |
+| `LetRec`    | `e0 With Mutually Recursive l1 = e1 ...` | `e0 With Mutually Recursve l1 = e1 ...` |
+
+In the new Postgres-style syntax, extra information will appear on the
+next line: for joins, it will be the join pipelines; for
```

Comment:
LIR has MFPs inside pipelines. We'll need to decide what to do with these. (MIR doesn't have these, so the current MIR EXPLAIN is often lying by saying that the MFP is executed after the join, when in fact the lowering pushes the MFP inside the pipelines.)

Maybe we can just leave off all the joins paths by default, unless `WITH (JOIN IMPLEMENTATIONS)` is specified. But the Filters (and also maps and projects if we generally decide to show them) will have to appear somewhere.

Reply:
I think we're going to want to see join order though. Postgres always shows join order (implicitly, via AST structure) along with the join conditions at each step of the join (`Hash Cond`, `Index Cond`).

Reply:
I'm still not sure how to show the MFPs that are pushed inside join paths. The current MIR EXPLAIN doesn't show these, which sometimes creates confusion, while the current EXPLAIN PHYSICAL PLAN shows these in an extremely verbose way.

> Postgres always shows join order (implicitly, via AST structure

This would get quite verbose for Delta joins, because we'd have a big tree for each join path. In other words, the verboseness of our printing would be quadratic with the number of join inputs.

Reply:
I think we could just show joins similarly to how current MIR EXPLAIN shows them, and think about this later.